multi batches upgrade  to fast storage

giskook opened the pull request:
When upgrade to fast IAVL. use one batch may cause OOM.so upgrade use multi batches.

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

giskook committed:
step upgrade

fix ut

use 10 for test celt2000

fix leak batch.close and add unittest

fix ut

add total statistics

add comments for commitGap and verboseGap, add log when delete stale fast nodes

more info about ndb.name

lyh169 commented on libs/iavl/nodedb.go:
```diff
@@ -791,8 +792,6 @@ func (ndb *nodeDB) Commit(batch dbm.Batch) error {
 		return errors.Wrap(err, "failed to write batch")
 	}
 
-	batch.Close()
```
should pay attention to the affect to other place

giskook committed:
add fss and create sub command

add cli fss command for upgrade to fast IAVL

add base folder

enable fast storage

add iavl log

add logs filter

add all modules to upgrade

fix nil map

add debug log

when upgrade to IAVL with store, open db every time

use distict var name with prefix db

only support acc and evm store to update to FSS

rename module keys to store keys

giskook committed:
add compact when upgrade to fast storage

full prefixdb's compact

lyh169 commented on libs/tm-db/types.go:
```diff
@@ -59,6 +59,9 @@ type DB interface {
 	Stats() map[string]string
+
+	// Compact compact the database
+	Compact() error
```
suggest `Compact()` with input param

giskook replied:
As we only use `Compact()` for upgrade to FSS now. we do not compact for specify range. so we do not use param

lyh169 commented on libs/iavl/mutable_tree.go:
```diff
@@ -658,6 +658,13 @@ func (tree *MutableTree) enableFastStorageAndCommitIfNotEnabled() (bool, error)
 	}
+
+	tree.log(IavlInfo, "Compacting IAVL...")
+	if err := tree.ndb.db.Compact(); err != nil {
```
suggest this `tree.ndb.db.Compact()` only work in the batch data save in db; other not  put to use

giskook replied:
Our compact occurs when we upgrade to FSS. so put the `Compact()` here.

lyh169 commented on cmd/exchaind/fss/create.go:
```diff
@@ -0,0 +1,79 @@
+	return []string{
+		auth.StoreKey,
+		evm.StoreKey,
+	}
```
`getStoreKeys()` is only two count ? or it can get from db?

giskook replied:
Because only `acc` and `evm` nodes much more than others and will upgrade take long time. So here only deal with the two store keys

giskook committed:
Merge branch 'dev' into zhangkai/upgrade

giskook committed:
Merge branch 'dev' into zhangkai/upgrade

KamiD committed:
Merge branch 'dev' into zhangkai/upgrade

giskook committed:
Merge branch 'dev' into zhangkai/upgrade

giskook committed:
Merge branch 'dev' into zhangkai/upgrade
